rllib/policy/tf_policy.py

```diff
@@ -330,18 +342,35 @@ def _initialize_loss(self, loss: TensorType,
             self._loss_input_dict["state_in_{}".format(i)] = ph
 
         if self.model and not isinstance(self.model, tf.keras.Model):
-            self._loss = self.model.custom_loss(loss, self._loss_input_dict)
+            self._losses = self.model.custom_loss(losses,
+                                                  self._loss_input_dict)
             self._stats_fetches.update({"model": self.model.metrics()})
         else:
-            self._loss = loss
-
-        if self._optimizer is None:
-            self._optimizer = self.optimizer()
-        self._grads_and_vars = [
-            (g, v) for (g, v) in self.gradients(self._optimizer, self._loss)
-            if g is not None
-        ]
-        self._grads = [g for (g, v) in self._grads_and_vars]
+            self._losses = losses
+        # Backward compatibility.
+        self._loss = self._losses[0] if self._losses is not None else None
+
+        if not self._optimizers:
+            self._optimizers = force_list(self.optimizer())
+            # Backward compatibility.
+            self._optimizer = self._optimizers[0] if self._optimizers else None
+
+        # Supporting more than one loss/optimizer.
+        if self.config["_tf_policy_handles_more_than_one_loss"]:
+            self._grads_and_vars = []
+            self._grads = []
+            for group in self.gradients(self._optimizers, self._losses):
+                g_and_v = [(g, v) for (g, v) in group if g is not None]
+                self._grads_and_vars.append(g_and_v)
+                self._grads.append([g for (g, v) in g_and_v])
+        # Only one optimizer and and loss term.
+        else:
+            self._grads_and_vars = [
+                (g, v)
+                for (g, v) in self.gradients(self._optimizer, self._loss)
```

Comment:
after seeing all these ifs, I wonder if the code will be simpler if we always use self._optimizers and self._losses here.
and handle things as single element array everywhere.

Reply:
Normally, yes, but the problem is those children of TFPolicy that don't have the `self._optimizers` property as they inherited directly from TFPolicy and did not build the class with the `build_tf_policy` utility function. :)

Reply:
We are (very carefully) introducing a new API here and have to make 100% sure, nothing breaks for users (and some of our algos) still using the old one. Once we can deprecate the old API, we can remove all these ugly if-blocks. But that'll take some time and some PRs introducing soft-warnings for users still on the old API.